File: android/lib/src/main/java/org/silvermine/downloadmanager/ProgressTracker.kt
package org.silvermine.downloadmanager

/**
 * Decides when a progress update should be emitted: on every change of integer
 * percent when [totalBytes] is known and positive, otherwise every
 * [BYTES_THRESHOLD] bytes.
 *
 * Tracks [receivedBytes] internally, so the download loop only calls [advance]
 * with each chunk size and checks [shouldEmit]. Mirrors the Rust
 * `ProgressTracker` in `crates/download-manager/src/downloader.rs`.
 */
internal class ProgressTracker(receivedBytes: Long, private val totalBytes: Long?) {
   var receivedBytes: Long = receivedBytes
      private set

   private var lastEmittedPercent: Long = percent(receivedBytes, totalBytes)
   private var lastEmittedBytes: Long = receivedBytes

   fun advance(bytesWritten: Long) {
      receivedBytes += bytesWritten
   }

   fun shouldEmit(): Boolean {
      if (totalBytes == null || totalBytes <= 0L) {
         return (receivedBytes - lastEmittedBytes) >= BYTES_THRESHOLD
      }

      // `percent >= 100` never emits — the call site suppresses it. Kept because a
      // true still drives the per-chunk store poll that detects pause and cancel.
      val percent = percent(receivedBytes, totalBytes)
      return (percent >= 100L || percent > lastEmittedPercent)
   }

   // The total is guarded as positive, not merely non-null: a wrapped Content-Length
   // sum would otherwise make every download read as complete from its first byte,
   // suppressing the emissions the loop gates behind this.
   fun isComplete(): Boolean = (totalBytes != null && totalBytes > 0L && receivedBytes >= totalBytes)

   fun markEmitted() {
      lastEmittedPercent = percent(receivedBytes, totalBytes)
      lastEmittedBytes = receivedBytes
   }

   companion object {
      const val BYTES_THRESHOLD = 1024L * 1024L

      private fun percent(bytes: Long, total: Long?): Long {
         if (total == null || total <= 0L) {
            return 0L
         }
         return ((bytes * 100L) / total)
      }
   }
}

File: ios/DownloadManagerKit/Sources/DownloadManagerKit/ProgressTracker.swift
//
//  ProgressTracker.swift
//  DownloadManagerKit
//

import Foundation

/// Decides when a progress update should be emitted: on every change of integer
/// percent when `totalBytes` is known and positive, otherwise every
/// `bytesThreshold` bytes.
///
/// Mirrors the emission rule of the Rust `ProgressTracker` in
/// `crates/download-manager/src/downloader.rs`. Rust accumulates chunk sizes as
/// it streams; URLSession reports a cumulative count per callback, so this
/// compares against the last count emitted instead of holding state.
struct ProgressTracker {
   static let bytesThreshold: UInt64 = 1024 * 1024

   private let receivedBytes: UInt64
   private let totalBytes: UInt64?
   private let lastEmittedBytes: UInt64

   init(lastEmittedBytes: UInt64, receivedBytes: UInt64, totalBytes: UInt64?) {
      self.receivedBytes = receivedBytes
      self.totalBytes = totalBytes
      self.lastEmittedBytes = lastEmittedBytes
   }

   var shouldEmit: Bool {
      // A restarted download (rejected resume data, or a server that stops
      // honouring Range) reports fewer bytes than the store last saw. Emit so
      // the baseline resets, rather than going quiet until it climbs back past
      // the old count. Also guards the unsigned subtraction below. Rust owns its
      // accumulator, so it never sees this.
      guard receivedBytes >= lastEmittedBytes else {
         return true
      }

      guard let total = totalBytes, total > 0 else {
         return (receivedBytes - lastEmittedBytes) >= Self.bytesThreshold
      }

      let percent = ((receivedBytes * 100) / total)
      let lastEmittedPercent = ((lastEmittedBytes * 100) / total)

      // `percent >= 100` never emits — the call site guards on it. Inert here, where
      // no store poll follows; kept so the three trackers read alike.
      return (percent >= 100 || percent > lastEmittedPercent)
   }

   var isComplete: Bool {
      // Positive, not merely non-nil — a zero total would otherwise read as
      // complete from the first byte. Mirrors the Android guard.
      guard let total = totalBytes, total > 0 else {
         return false
      }
      return receivedBytes >= total
   }
}
